### PR TITLE
opt: consider limit hint when costing distinct-on expressions

### PR DESCRIPTION
**Pull request by mgartner**
#### opt: add coster tests for disctinct-on expressions

This commit adds coster tests for distinct-on expressions. There were no
such tests before.

Release note: None

#### opt: consider limit hint when costing distinct-on expressions

Limit hints are now considered when costing DistinctOn expressions,
similarly to streaming GroupBy expressions. This matches the behavior of
the execution logic, where a DistinctOn emits a row with distinct
grouping columns as soon as it is read from its input. It does not need
to read all of its input to begin emitting rows.

Release note (performance improvement): The optimizer now costs
distinct-on operators more accurately. It may produce more efficient
query plans in some cases.

Epic: CRDB-37714

#### opt: add optimizer_use_improved_distinct_on_limit_hint_costing

The `optimizer_use_improved_distinct_on_limit_hint_costing` session
setting has been added which enables the improved costing of DistinctOn
expressions with limit hints added in the previous commit. It is enabled
by default.

Release note: None

### CODE REVIEW

**Review by DrewKimball — APPROVED** [review]
:lgtm: Nice work!

> Reviewed 1 of 1 files at r1, 3 of 3 files at r2, 12 of 12 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/122442#-:-NvbzsDFAx-t5q1RPUd-:bcmw7ss)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @michae2)

File: pkg/sql/opt/xform/coster.go [hunk 1]
```diff
@@ -1405,12 +1405,18 @@ func (c *coster) computeGroupingCost(grouping memo.RelExpr, required *physical.R
...
-	if (streamingType != memo.NoStreaming) && grouping.Op() == opt.GroupByOp && required.LimitHint > 0 {
-		inputRowCount = streamingGroupByInputLimitHint(inputRowCount, outputRowCount, required.LimitHint)
-		outputRowCount = math.Min(outputRowCount, required.LimitHint)
+	if required.LimitHint > 0 {
+		if grouping.Op() == opt.GroupByOp && streamingType != memo.NoStreaming {
+			inputRowCount = streamingGroupByInputLimitHint(inputRowCount, outputRowCount, required.LimitHint)
+			outputRowCount = math.Min(outputRowCount, required.LimitHint)
+		} else if grouping.Op() == opt.DistinctOnOp &&
+			c.evalCtx.SessionData().OptimizerUseImprovedDistinctOnLimitHintCosting {
+			inputRowCount = distinctOnLimitHint(outputRowCount, required.LimitHint)
```
Comment (mgartner) [inline pkg/sql/opt/xform/coster.go:1417]: Note: We might need to use `streamingGroupByInputLimitHint` for streaming distinct-on expressions, to make the plan changes go away.

**Comment by mgartner** [comment]
There are a few intriguing plan changes, mostly from streaming distinct-ons to hash distinct-ons. @DrewKimball and I spent some time looking at this and we aren't super concerned with the changes—they should have very little impact on performance, if any. https://github.com/cockroachdb/cockroach/pull/122644 tracks some potential follow-up work.

**Pull request merged**